mp4parse/src/lib.rs

```diff
@@ -916,6 +1173,338 @@ fn skip_box_remain<T: Read>(src: &mut BMFFBox<T>) -> Result<()> {
     skip(src, remain)
 }
 
+/// Read the contents of an AVIF file
+///
+/// Metadata is accumulated in the passed-through `AvifContext` struct,
+/// which can be examined later.
+pub fn read_avif<T: Read>(f: &mut T, context: &mut AvifContext) -> Result<()> {
+    let mut f = OffsetReader::new(f);
+
+    let mut iter = BoxIter::new(&mut f);
+
+    // 'ftyp' box must occur first; see ISO 14496-12:2015 § 4.3.1
+    if let Some(mut b) = iter.next_box()? {
+        if b.head.name == BoxType::FileTypeBox {
+            let ftyp = read_ftyp(&mut b)?;
+            if !ftyp.compatible_brands.contains(&FourCC::from("mif1")) {
+                return Err(Error::InvalidData("compatible_brands must contain 'mif1'"));
+            }
+        } else {
+            return Err(Error::InvalidData("'ftyp' box must occur first"))
+        }
+    }
+
+    let mut read_meta = false;
+    let mut mdats = vec![];
+    let mut primary_item_extents = None;
+    let mut primary_item_extents_data: Vec<Vec<u8>> = vec![];
+
+    while let Some(mut b) = iter.next_box()? {
+        match b.head.name {
+            BoxType::MetadataBox => {
+                if read_meta {
+                    return Err(Error::InvalidData("There should be zero or one meta boxes per ISO 14496-12:2015 § 8.11.1.1"));
+                }
+                read_meta = true;
+                let primary_item_loc = read_avif_meta(&mut b)?;
+                match primary_item_loc.construction_method {
+                    ConstructionMethod::FileOffset => {
+                        primary_item_extents = Some(primary_item_loc.extents);
+                        primary_item_extents_data = primary_item_extents.iter().map(|_| vec![]).collect();
+                    }
+                    _ => return Err(Error::Unsupported("unsupported construction_method")),
+                }
+            }
+            BoxType::MediaDataBox => { // See ISO 14496-12:2015 § 8.1.1
+                // If we know our primary item location by this point, try to read it out of this
+                // mdat directly and avoid a copy
+                if let Some(extents) = &primary_item_extents {
+                    for (extent, data) in extents.iter().zip(primary_item_extents_data.iter_mut()) {
+                        if b.contains_extent(&extent.extent_range) {
+                            b.read_extent(&extent.extent_range, data)?;
+                        }
+                    }
+                }
+
+                // Store any remaining data for potential later extraction
+                if b.bytes_left() > 0 {
+                    let offset = b.offset();
+                    let mut data = Vec::with_capacity(b.bytes_left().try_into()?);
```

Comment:
Why does [`allocate_read_buf`](allocate_read_buf) explicitly zero the memory as opposed to just allocating the additional capacity? For [the single caller](https://github.com/mozilla/mp4parse-rust/blob/bbb4a1217d2826e95b0a843e73c9e1a2faa23953/mp4parse/src/lib.rs#L2726), it seems to incur an unnecessary cost.

I have other `Vec::with_capacity` callers that aren't `Vec<u8>`, so I'll need to make some change here. What's your preference?
1. Modify `allocate_read_buf` to take a type parameter and not fill its capacity with zeroes
1. Leave `allocate_read_buf` as is, but add a new function like
    ```rust
    fn vec_with_capacity<T>(capacity: usize) -> std::result::Result<Vec<T>, ()> {
        #[cfg(feature = "mp4parse_fallible")]
        {
            let mut v = Vec::new();
            FallibleVec::try_reserve(&mut v, capacity)?;
            Ok(v)
        }
        #[cfg(not(feature = "mp4parse_fallible"))]
        {
            Ok(Vec::with_capacity(capacity))
        }
    }
    ```
1. Something else
?

Reply:
The zeroing is due to https://github.com/mozilla/mp4parse-rust/issues/172

Reply:
Ooh, that's subtle. I'm going to add a comment back to the issue for posterity, if you don't mind.

Do you have any preference regarding whether or not to add `vec_with_capacity` versus modifying `allocate_read_buf` to handle types other than `u8`?

If I implement `vec_with_capacity` as described above, I don't think we need to worry about the issue in https://github.com/mozilla/mp4parse-rust/issues/172 since it doesn't provide any access to potentially uninitialized memory the way [the old version of `allocate_read_buf`](https://github.com/mozilla/mp4parse-rust/blob/26e614bbcb4d7322dc2e4b7e15391bdb30b9f7be/mp4parse/src/lib.rs#L70) did. It gets around that by using [`Read::read_to_end`](https://doc.rust-lang.org/stable/std/io/trait.Read.html#method.read_to_end) which takes a `&mut Vec<u8>` `buf` rather than [`Read::read`](https://doc.rust-lang.org/stable/std/io/trait.Read.html#tymethod.read), which takes a `&mut [u8]` `buf`.

It occurs to me that we could eliminate the need for zeroing out the memory in `allocate_read_buf` (and for `allocate_read_buf` ifself) if we switched [the code here](https://github.com/mozilla/mp4parse-rust/blob/bbb4a1217d2826e95b0a843e73c9e1a2faa23953/mp4parse/src/lib.rs#L2726-L2727) from using `read` to `read_to_end` with `take` like so:

```rust
    if let Ok(mut buf) = allocate_read_buf(size) {
        let r = src.read(&mut buf)?;
```
becomes
```rust
    if let Ok(mut buf) = vec_with_capacity(size) {
        let r = src.take(size.to_u64()).read_to_end(&mut buf)?;
```

The only caveat I can see is that if the `Vec` passed to `read_to_end` didn't already have sufficient capacity, it could end up doing infallible allocation. But that shouldn't be an issue as long as we use `vec_with_capacity`.

Thoughts?

Reply:
> Ooh, that's subtle. I'm going to add a comment back to the issue for posterity, if you don't mind.

Good idea, thanks.

> Do you have any preference regarding whether or not to add `vec_with_capacity` versus modifying `allocate_read_buf` to handle types other than `u8`?

I think we only need one, so making it handle other simple types makes sense.  I don't mind which one we keep but vec_with_capacity is probably a better name.

> If I implement `vec_with_capacity` as described above, I don't think we need to worry about the issue in #172 since it doesn't provide any access to potentially uninitialized memory the way [the old version of `allocate_read_buf`](https://github.com/mozilla/mp4parse-rust/blob/26e614bbcb4d7322dc2e4b7e15391bdb30b9f7be/mp4parse/src/lib.rs#L70) did. It gets around that by using [`Read::read_to_end`](https://doc.rust-lang.org/stable/std/io/trait.Read.html#method.read_to_end) which takes a `&mut Vec<u8>` `buf` rather than [`Read::read`](https://doc.rust-lang.org/stable/std/io/trait.Read.html#tymethod.read), which takes a `&mut [u8]` `buf`.

I think you still have the same issue, in that you can't pass a Vec with an uninitialized backing store to Read::read_to_end because the trait doesn't promise implementations won't read from the &mut Vec<u8> before writing to it.  Although the docs only include a warning for Read::read, it's actually a restriction on the entire trait.  Once Read::initializer is stable, we can use that to signal that our implementation of Read doesn't read from the buffers passed in.

Reply:
Maybe I'm not understanding https://github.com/mozilla/mp4parse-rust/issues/172 correctly. Can you let me know where my logic is flawed?

As I understand it, the unsoundness came from [this line in the old version of `allocate_read_buf`](https://github.com/mozilla/mp4parse-rust/blob/26e614bbcb4d7322dc2e4b7e15391bdb30b9f7be/mp4parse/src/lib.rs#L75):
```rust
        unsafe { buf.set_len(size); }
```
because it violated [the safety requirement of `Vec::set_len()`](https://doc.rust-lang.org/stable/std/vec/struct.Vec.html?search=#safety-1):
> The elements at `old_len..new_len` must be initialized.

And as mentioned in [`Read::read`](https://doc.rust-lang.org/stable/std/io/trait.Read.html#tymethod.read):
> It is your responsibility to make sure that `buf` is initialized before calling `read`. Calling `read` with an uninitialized `buf` (of the kind one obtains via `MaybeUninit<T>`) is not safe, and can lead to undefined behavior.

So, because `allocate_read_buf` called `set_len` without initializing the additional capacity, it was uninitialized and the call to `Read::read` was unsound. More generally, `allocate_read_buf` returned a vector that was invalid because uninitialized memory was now accessible via the safe interface.

Hovever, I don't see how the `vec_with_capacity` case has the same issue. [There's no `unsafe` code in its implementation](https://github.com/mozilla/mp4parse-rust/blob/cd8b3d7c6663a8866ab4a891749e54b809257ebf/mp4parse/src/lib.rs#L144-L146), so how does it introduce any uninitialized memory into the memory of the vector's elements? The only potentially uninitialized memory should be in the internal buffer beyond the valid length of the vector.

Take [this example code](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=060c6f69872ebd39e69d585effc49063):

```rust
    let mut reader: &[u8] = b"1234567890";
    let mut buf = Vec::with_capacity(5);
    reader.read(&mut buf).unwrap();
```

After the first call to `reader.read`, `buf` will be empy. Nothing will be read into it because despite having a capacity of 10, it has a length of 0. Contrast that with [identical code except replacing `read` with `read_to_end`](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=576c38ad9c0555df7ac4bfcc33587695), and the result is that `buf` contains all 10 elements.

I think the difference is easy to miss because in both cases we pass `&mut buf`, but in the `read` call, that is auto derefed to a `&mut [u8]` slice, and in the `read_to_end` case, it remains a `&mut Vec<u8>`. The callee can't change the length of the prior, but it can change the length of the latter, which is why [`Read::read_to_end`](https://doc.rust-lang.org/stable/std/io/trait.Read.html#method.read_to_end)'s documentation describes it as "append[ing] data to `buf`, whereas [`Read::read`](https://doc.rust-lang.org/stable/std/io/trait.Read.html#tymethod.read)'s says it "pull[s] some bytes from this source into the specified buffer".

Unless I'm missing something, we have the opportunity to avoid the cost of zeroing out the buffer (as with the earlier version of `allocate_read_buf`), but eliminating the potential for unsoundness. Does that seem worth it?

Reply:
I'm confusing things by thinking only about the allocate_read_buf impl. I agree with your comment (thanks for the detailed reasoning!), so replacing  allocate_read_buf impl + read with vec_with_capacity + read_to_end allows us to eliminate the extra initialization safely.

Reply:
I ended up taking a slightly different approach from what I suggested in https://github.com/mozilla/mp4parse-rust/pull/193#discussion_r386771240 since I realized I could simplify things by making [our new `read_to_end` with fallible allocation support](https://github.com/mozilla/mp4parse-rust/blob/avif-pitm/mp4parse/src/lib.rs#L168) take care of doing all the allocation.

The heart of the change is [here](https://github.com/mozilla/mp4parse-rust/pull/193/commits/9246ad69e75f93644816b9a31ac3c90eb20cd92d). I ended up splitting out some other cleanup-type changes into [the previous commit](https://github.com/mozilla/mp4parse-rust/pull/193/commits/ba1b7c94c9d2d2e7f39019b9ea9cbc532b810d90).

@kinetiknz, @SingingTree: could you please give this one more look before I merge?